### PR TITLE
Add channel limit per node

### PR DESCRIPTION
**Pull request by SimonUnge**
## Proposed Changes

A per node channel limit.

An opt in setting to set a hard max channel limit, per node, regardless of channel max per connection.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [x] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [x] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.

### CODE REVIEW

File: deps/rabbit/src/rabbit_reader.erl [hunk 1]
```diff
@@ -915,11 +915,45 @@ create_channel(Channel,
...
+            {true, Limit, Fmt}
+    end.
+
+is_over_node_channel_limit() ->
+    case rabbit_misc:get_env(rabbit, channel_max_per_node, 0) of
+        0 ->
+            false;
+        NodeLimit ->
+            %% Only fetch this if a limit is set
+            CurrNodeChannels = length(rabbit_channel_tracking:list_on_node(node())),
```
Comment (dcorbacho) [inline deps/rabbit/src/rabbit_reader.erl:950]: I'd suggest to implement a new function in `rabbit_channel_tracking` that gives you the number of channels. `ets:info(Tab, size)` gives you the size of the table without having to list all elements. With many channels, it's going to be a cheaper call

Comment (SimonUnge) [reply deps/rabbit/src/rabbit_reader.erl:950]: Ah, smart will do!

Comment (SimonUnge) [reply deps/rabbit/src/rabbit_reader.erl]: Done @dcorbacho

**Comment by michaelklishin** [comment]
This was a rebase.

File: deps/rabbit/priv/schema/rabbit.schema [hunk 2]
```diff
@@ -889,6 +889,9 @@ end}.
 
 {mapping, "channel_max", "rabbit.channel_max", [{datatype, integer}]}.
 
+{mapping, "channel_max_per_node", "rabbit.channel_max_per_node",
+          [{datatype, integer}, {validators, ["non_negative_integer"]}]}.
```
Comment (michaelklishin) [inline deps/rabbit/priv/schema/rabbit.schema:893]: This should be a positive integer, setting this value to 0 does not make much sense (unless you want to effectively block AMQP 0-9-1 clients specifically, which I don't think is what we're after here).

Comment (SimonUnge) [reply deps/rabbit/priv/schema/rabbit.schema:893]: Good point. Actually, 0 here means 'infinite', so I might just change that to be the default for readability.

Comment (michaelklishin) [reply deps/rabbit/priv/schema/rabbit.schema:893]: Ah, right. Then this validator makes sense. As for the default, we use `cuttlefish:unset/0` as a return value when we want to default in the code to kick in.

Comment (SimonUnge) [reply deps/rabbit/priv/schema/rabbit.schema]: Done, changed 0 -> infinity

Comment (michaelklishin) [reply deps/rabbit/priv/schema/rabbit.schema]: OK, I think it makes more sense than `0`, and for new settings we can adopt this new value. Most users who want no limits probably won't set it in the first place.

**Review by michaelklishin — APPROVED** [review]

**Pull request merged**